hudi-utilities/src/main/java/org/apache/hudi/utilities/keygen/MultiFormatTimestampBasedKeyGenerator.java

```diff
@@ -0,0 +1,183 @@
+import org.joda.time.format.DateTimeFormatterBuilder;
+import org.joda.time.format.DateTimeParser;
+
+/**
+ * Key generator, that relies on timestamps for partitioning field. Still picks record key by name.
+ */
+public class MultiFormatTimestampBasedKeyGenerator extends SimpleKeyGenerator {
+
+  enum TimestampType implements Serializable {
+    UNIX_TIMESTAMP, DATE_STRING, MIXED, EPOCHMILLISECONDS
+  }
+
+  private final TimestampType timestampType;
+  private final String outputDateFormat;
+  private final String configInputDateFormatList;
+  private final String configInputDateFormatDelimiter;
+
+
+  // TimeZone detailed settings reference
+  // https://docs.oracle.com/javase/8/docs/api/java/util/TimeZone.html
+  private final DateTimeZone inputDateTimeZone;
+  private final DateTimeZone outputDateTimeZone;
+
+  /**
+   * Supported configs.
+   */
+  static class Config {
+    // One value from TimestampType above
+    private static final String TIMESTAMP_TYPE_FIELD_PROP                             = "hoodie.deltastreamer.keygen.timebased.timestamp.type";
+
+    private static final String TIMESTAMP_INPUT_DATE_FORMAT_LIST_PROP                 = "hoodie.deltastreamer.keygen.timebased.input.dateformatlist";
+    private static final String TIMESTAMP_INPUT_DATE_FORMAT_LIST_DELIMETER_REGEX_PROP = "hoodie.deltastreamer.keygen.timebased.input.dateformatlistdelimiterregex";
+    private static final String TIMESTAMP_INPUT_TIMEZONE_FORMAT_PROP                  = "hoodie.deltastreamer.keygen.timebased.input.timezone";
+
+    private static final String TIMESTAMP_OUTPUT_DATE_FORMAT_PROP                     = "hoodie.deltastreamer.keygen.timebased.output.dateformat";
+    private static final String TIMESTAMP_OUTPUT_TIMEZONE_FORMAT_PROP                 = "hoodie.deltastreamer.keygen.timebased.output.timezone";
+  }
+
+  public MultiFormatTimestampBasedKeyGenerator(TypedProperties config) {
+    super(config);
+    DataSourceUtils.checkRequiredProperties(config,
+        Arrays.asList(Config.TIMESTAMP_TYPE_FIELD_PROP, Config.TIMESTAMP_OUTPUT_DATE_FORMAT_PROP));
+
+    String inputTimeZone = config.getString(Config.TIMESTAMP_INPUT_TIMEZONE_FORMAT_PROP, "");
+    String outputTimeZone = config.getString(Config.TIMESTAMP_OUTPUT_TIMEZONE_FORMAT_PROP, "");
+
+    this.timestampType = TimestampType.valueOf(config.getString(Config.TIMESTAMP_TYPE_FIELD_PROP));
+    this.outputDateFormat = config.getString(Config.TIMESTAMP_OUTPUT_DATE_FORMAT_PROP);
+    this.configInputDateFormatList = config.getString(Config.TIMESTAMP_INPUT_DATE_FORMAT_LIST_PROP, "");
+
+    String inputDateFormatDelimiter = this.config.getString(Config.TIMESTAMP_INPUT_DATE_FORMAT_LIST_DELIMETER_REGEX_PROP, ",").trim();
+    inputDateFormatDelimiter = inputDateFormatDelimiter.isEmpty() ? "," : inputDateFormatDelimiter;
+    this.configInputDateFormatDelimiter = inputDateFormatDelimiter;
+
+    if (inputTimeZone != null && !inputTimeZone.trim().isEmpty()) {
```

Comment:
`inputTimeZone` wont be null at all since `""` is the default value?

Reply:
Yup, let me take a look at that - it may be a remnant of my iterations through failures at runtime.

Reply:
Oh, btw @vinothchandar - on the existing timestampgenerator - I too thought about extending that thing to make it happen, but while I was researching how I wanted to approach this, using the SimpleDateFormat class that the TimestampBasedKeyGenerator means the options would have been to create a list of formats the user wanted to parse, and then basically do try catch iterations around each provided formatter (which is an expensive process), and then decide to throw at the end if none matched.  So, the primary reason I went the Joda time route is it has a built in ability to handle and attempt multiple formats, supposedly in a more performant way.  I have not taken the time to verify whether it's faster or not, but with the existing time classes being used by the existing timestamp key generator, the options were limited.  

Just wanted to throw that out there to give some background information on why I chose to split to a new class rather than extend the old - basically extending the old would likely have been rewriting the old (which may have been the right path).